File: CSharp/Builders/BuildANewThing.cs
using System;

namespace ThingModel.Builders
{
	public class BuildANewThing
	{
		private ThingType _type;

		private BuildANewThing(ThingType type)
		{
			_type = type;
		}


		public ThingPropertyBuilder IdentifiedBy(String id)
		{
			return new ThingPropertyBuilder(new Thing(id, _type));
		}

		public static BuildANewThing WithoutType()
		{
			return new BuildANewThing(null);
		}

		public static BuildANewThing As(ThingType type)
		{
			return new BuildANewThing(type);
		}

		public class ThingPropertyBuilder
		{
			private readonly Thing _thing;

			public ThingPropertyBuilder(Thing thing)
			{
				_thing = thing;
			}

			public ThingPropertyBuilder ContainingA
			{
				get { return this; }
			}

			public ThingPropertyBuilder ContainingAn
			{
				get { return this; }
			}

			public ThingPropertyBuilder AndA
			{
				get { return this; }
			}

			public ThingPropertyBuilder AndAn
			{
				get { return this; }
			}

			public ThingPropertyBuilder String(String key, String value)
			{
				_thing.SetProperty(new Property.String(key, value));
				return this;
			}

			public ThingPropertyBuilder Location(Location.Equatorial value)
			{
				return Location("location", value);
			}

			public ThingPropertyBuilder Location(String key, Location.Equatorial value)
			{
				_thing.SetProperty(new Property.Location.Equatorial(key, value));
				return this;
			}

			public ThingPropertyBuilder Location(Location.LatLng value)
			{
				return Location("location", value);
			}

			public ThingPropertyBuilder Location(String key, Location.LatLng value)
			{
				_thing.SetProperty(new Property.Location.LatLng(key, value));
				return this;
			}

			public ThingPropertyBuilder Location(Location.Point value)
			{
				return Location("location", value);
			}

			public ThingPropertyBuilder Location(String key, Location.Point value)
			{
				_thing.SetProperty(new Property.Location.Point(key, value));
				return this;
			}

			public ThingPropertyBuilder Double(String key, double value)
			{
				_thing.SetProperty(new Property.Double(key, value));
				return this;
			}

			public ThingPropertyBuilder Int(String key, int value)
			{
				_thing.SetProperty(new Property.Int(key, value));
				return this;
			}

			public ThingPropertyBuilder Boolean(String key, bool value)
			{
				_thing.SetProperty(new Property.Boolean(key, value));
				return this;
			}

			public ThingPropertyBuilder DateTime(String key, DateTime value)
			{
				_thing.SetProperty(new Property.DateTime(key, value));
				return this;
			}

			public static implicit operator Thing(ThingPropertyBuilder b)
			{
				return b._thing;
			}
		}
	}
}

File: TypeScript/Builders/BuildANewThing.ts
 module ThingModel {
	 export class BuildANewThing {
		 private type: ThingType;

		 /* private */ constructor(type: ThingType) {
			 this.type = type;
		 }

		 public static get WithoutType(): BuildANewThing {
			 return new BuildANewThing(null);
		 }

		 public static As(type : ThingType): BuildANewThing {
			 return new BuildANewThing(type);
		 }

		 public IdentifiedBy(id: string) : ThingPropertyBuilder {
			 return new ThingPropertyBuilder(new Thing(id, this.type));
		 }
	 }

	 export class ThingPropertyBuilder {
		 private thing: Thing;
		 public ContainingA : ThingPropertyBuilder;
		 public ContainingAn : ThingPropertyBuilder;
		 public AndA : ThingPropertyBuilder;
		 public AndAn : ThingPropertyBuilder;

		 constructor(thing: Thing) {
			 this.thing = thing;
			 this.ContainingA = this;
			 this.ContainingAn = this;
			 this.AndA = this;
			 this.AndAn = this;
		 }

		 public String(key: string, value: string): ThingPropertyBuilder {
			 this.thing.SetProperty(new Property.String(key, value));
			 return this;
		 }

		 public Double(key: string, value: number): ThingPropertyBuilder {
			 this.thing.SetProperty(new Property.Double(key, value));
			 return this;
		 }

		 public Int(key: string, value: number): ThingPropertyBuilder {
			 this.thing.SetProperty(new Property.Int(key, value));
			 return this;
		 }

		 public Boolean(key: string, value: boolean): ThingPropertyBuilder {
			 this.thing.SetProperty(new Property.Boolean(key, value));
			 return this;
		 }

		 public DateTime(key: string, value: Date): ThingPropertyBuilder {
			 this.thing.SetProperty(new Property.DateTime(key, value));
			 return this;
		 }

		 public Location(value: Location):	 ThingPropertyBuilder;
		 public Location(key: string, value: Location): ThingPropertyBuilder;
		 public Location(mixed: any, value?: Location): ThingPropertyBuilder {

			 var key;
			 if (value) {
				 key = mixed;
			 } else {
				 key = "location";
				 value = mixed;
			 }

			 switch(value.type) {
				 case "equatorial":
					this.thing.SetProperty(new Property.Location.Equatorial(key, <Location.Equatorial>value));
					break;
				 case "latlng":
					this.thing.SetProperty(new Property.Location.LatLng(key,<Location.LatLng>value));
					 break;
				 case "point":
				default :
					this.thing.SetProperty(new Property.Location.Point(key, <Location.Point>value));
					break;

			 }
			 return this;
		 }

		 public Build(): Thing {
			 return this.thing;
		 }
	 }
 }